include/networkit/graph/GraphTools.hpp

```diff
@@ -97,6 +184,146 @@ Graph getRemappedGraph(const Graph& graph, count numNodes,
     return Gnew;
 }
 
+/**
+ * Calls the given BFS handle with distance parameter.
+ */
+template <class F>
+auto callBFSHandle(F &f, node u, count dist) -> decltype(f(u, dist)) {
+    return f(u, dist);
+}
+
+/**
+ * Calls the given BFS handle without distance parameter.
+ */
+template <class F>
+auto callBFSHandle(F &f, node u, count) -> decltype(f(u)) {
+    return f(u);
+}
+
+/**
+ * Iterate over nodes in breadth-first search order starting from the nodes in the given range.
+ *
+ * @param G The input graph.
+ * @param first Start of the range.
+ * @param last End of the range.
+ * @param handle Callback function, takes one parameter <code>(node)</code>.
+ */
+template <class InputIt, typename L>
+void BFSfrom(const Graph &G, InputIt first, InputIt last, L handle) {
+    std::vector<bool> visited(G.upperNodeIdBound());
+    std::queue<node> q, qNext;
+    count dist = 0;
+
+    // enqueue start nodes
+    for (; first != last; ++first) {
+        q.push(*first);
+        visited[*first] = true;
+    }
+
+    do {
+        auto u = q.front();
+        q.pop();
+        // apply function
+        callBFSHandle(handle, u, dist);
+        G.forNeighborsOf(u, [&](node v) {
+            if (!visited[v]) {
+                qNext.push(v);
+                visited[v] = true;
+            }
+        });
+
+        if (q.empty() && !qNext.empty()) {
+            q.swap(qNext);
+            ++dist;
+        }
+    } while (!q.empty());
+}
+
+/**
+ * Iterate over the nodes of the input graph in breadth-first search order starting from @a source
+ * until the connected component of @a source has been visited.
+ *
+ * @param G The input graph.
+ * @param source Source node of the breadth-first search.
+ * @param handle Callback function, takes one parameter <code>(node)</code>.
+ */
+template <typename L> void BFSfrom(const Graph &G, node source, L handle) {
```

Comment:
Should we put BFS and DFS into dedicated headers (e.g. `graph/DFS.hpp` and `graph/BFS.hpp`)? To me it seems that they are build-blocks with a well-known name and somehow do not fit into the 'heap of algorithms that didn't fit in somewhere else'

Reply:
Also, if we really try to avoid the callback idiom in the future, we might have to add BFS/DFS iterators, which will unnecessarily blow up GraphTools.

Reply:
That makes sense, we might want to use a different name since `distance/BFS.hpp` already exist.

Reply:
So I moved BFS and DFS in separated headers, and defined a `GraphTraversal` namespace.